extensions/positron-reticulate/src/extension.ts

```diff
@@ -484,20 +484,45 @@ class ReticulateRuntimeSession implements positron.LanguageRuntimeSession {
 		// tied to the R session.
 		// We have to send a restart to the R session, and send a reticulate::repl_python()
 		// command to it.
-		this.pythonSession.shutdown(positron.RuntimeExitReason.Restart);
-		await this.rSession.restart();
-		const rSession = await getRSession();
-		rSession.execute(
-			'reticulate::repl_python()',
-			'start-reticulate',
-			positron.RuntimeCodeExecutionMode.Interactive,
-			positron.RuntimeErrorBehavior.Continue
+		const restart = await positron.window.showSimpleModalDialogPrompt(
+			'Restarting reticulate',
+			'This is will also restart the parent R session. Are you sure you want to continue?',
+			'Yes',
+			'No'
 		);
+
+		if (!restart) {
+			return;
+		}
+
+		// The events below will make sure that things occure in the right order:
+		// 1. shutdown the current reticulate session
+		// 2. restart the attached R session
+		// 3. start a new reticulate session.
+		this.pythonSession.onDidEndSession((sess) => {
+			this.rSession.restart();
+		});
+
+		const disposeListener = this.rSession.onDidChangeRuntimeState(async (e) => {
+			if (e === positron.RuntimeState.Ready) {
+				this.rSession.execute(
+					'reticulate::repl_python()',
+					'start-reticulate',
+					positron.RuntimeCodeExecutionMode.Interactive,
+					positron.RuntimeErrorBehavior.Continue
+				);
+				// This should only happen once, so we dispose of this event as soon
+				// as we have started reticulate.
+				disposeListener.dispose();
+			}
+		});
+
+		await this.shutdown(positron.RuntimeExitReason.SwitchRuntime);
```

Comment:
Should this be ExitReason.Restart?

Reply:
If we send a `ExitReason.Restart` the Jupyter Adapter will try to reinitialize it as soon as it exitted, but we actually want to wait for the new R session to be started.

https://github.com/posit-dev/positron/blob/4a849834de7c81aa0e14f511267664cf343c394b/extensions/jupyter-adapter/src/LanguageRuntimeAdapter.ts#L901-L914

My reasonming was that this 'looks like' we are effectively switching runtimes, as we don't preserve the session RuntimeSessionObject, etc. Maybe `RuntimeExitReason.Shutdown` is better?